File: L03-Time-Complexity/TapeEquilibrium.java
// Date: Nov 26, 2021
// https://app.codility.com/programmers/lessons/3-time_complexity/tape_equilibrium/

// use Sliding Window Technique
// time complexity: O(n)
// space complexity: O(1)
// result: https://app.codility.com/demo/results/trainingEXK2ZS-7XY/
public int solution(int[] A) {
	int N = A.length;
	int leftSum = A[0], rightSum = 0;
	for (int i = 1; i < N; i++) {
		rightSum += A[i];
	}
	int ans = Math.abs(leftSum - rightSum);
	for (int i = 1; i < N - 1; i++) {
		leftSum += A[i];
		rightSum -= A[i];
		ans = Math.min(ans, Math.abs(leftSum - rightSum));
	}
	return ans;
}

File: L03-Time-Complexity/TapeEquilibrium.swift
// Notice: 0 < P < N => each part has at least 1 element
// result: https://app.codility.com/demo/results/trainingDZWSW3-6BC/
public class TapeEquilibrium {
    public static func solution(_ A : inout [Int]) -> Int {
        var leftSum = A[0], rightSum = A[1..<A.count].reduce(0, +)
        // print("initial value of leftSum is \(leftSum)")
        // print("initial value of rightSum is \(rightSum)")
        var answer = abs(leftSum - rightSum)
        // print("initial value of answer is \(answer)")
        for number in A[1..<A.count - 1] {
            leftSum += number
            // print("leftSum is updated to \(leftSum)")
            rightSum -= number
            // print("rightSum is updated to \(rightSum)")
            answer = min(answer, abs(leftSum - rightSum))
        }
        return answer
    }
}
